crates/holochain/src/core/workflow/app_validation_workflow/validation_dependencies.rs

```diff
@@ -0,0 +1,296 @@
+use std::{
+    collections::{hash_map::Entry, HashMap, HashSet},
+    time::{Duration, Instant},
+};
+
+use holo_hash::{AnyDhtHash, DhtOpHash};
+use holochain_types::dht_op::DhtOpHashed;
+
+#[derive(Debug)]
+pub(super) struct MissingHashProperties {
+    depending_ops: HashSet<DhtOpHash>,
+    when_fetched: Instant,
+}
+
+/// In-memory struct to keep track of missing DHT hashes, which DhtOp depends on them
+/// and when the fetches for them were most recently tried.
+pub struct ValidationDependencies {
+    /// Missing hashes that are being fetched, along with a set of DhtOps that depend
+    /// on the hash and the timestamp a fetch was attempted.
+    pub(super) missing_hashes: HashMap<AnyDhtHash, MissingHashProperties>,
+}
+
+impl Default for ValidationDependencies {
+    fn default() -> Self {
+        ValidationDependencies::new()
+    }
+}
+
+impl ValidationDependencies {
+    const FETCH_TIMEOUT: Duration = Duration::from_secs(60);
+
+    pub fn new() -> Self {
+        Self {
+            missing_hashes: HashMap::new(),
+        }
+    }
+
+    /// Returns true if this is a new missing hash.
+    pub fn insert_missing_hash_for_op(&mut self, hash: AnyDhtHash, dht_op_hash: DhtOpHash) -> bool {
+        match self.missing_hashes.entry(hash.clone()) {
+            Entry::Vacant(entry) => {
+                let mut dht_op_hashes = HashSet::new();
+                dht_op_hashes.insert(dht_op_hash);
+                entry.insert(MissingHashProperties {
+                    depending_ops: dht_op_hashes,
+                    when_fetched: Instant::now(),
+                });
+                true
+            }
+            Entry::Occupied(mut entry) => {
+                entry.get_mut().depending_ops.insert(dht_op_hash.clone());
+                false
+            }
+        }
+    }
+
+    /// Iterates over given missing hashes and adds them to the internal collection if they are not present.
+    /// A collection of new missing hashes is returned.
+    /// Takes missing hashes and the depending op as parameters.
+    pub fn insert_missing_hashes_for_op(
```

Comment:
I'm still not perfectly happy with this fn name because it reflects only the side-effect and not the filtering aspect of it